src/template.c

```diff
@@ -5956,22 +5956,20 @@ void TemplateInstance::semantic(Scope *sc, Expressions *fargs)
         this->tnext = inst->tnext;
         inst->tnext = this;
 
-        // If the first instantiation was in speculative context, but this is not:
-        if (tinst && !inst->tinst && !inst->minst)
+        if (minst && minst->isRoot() && !(inst->minst && inst->minst->isRoot()))
```

Comment:
If a template is instantiated in speculative context, the context should handle *all* errors during the instantiation - it's including the errors in function body.

```d
void foo(T)(T t) { auto x = undefined_name; }
static assert(!is(typeof(foo(1))));   // the undefined error should be captured in here.
```

For that, all of speculative instantiations will be handled as root instances, and then, their semantic3 pass will surely get called.